### PR TITLE
Release of Stop and Copy and Mark and Sweep Garbage Collection

### PR DESCRIPTION
**Pull request by violinyap**
# Description

This PR includes the implementation of CS4215 project, Memory Management Visualiser.
No additional dependency required.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

- [ ] Use a test program, released in source-program*

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Anonymxtrix** [comment]
Hi @violinyap can I trouble you to rebase `violinyap:master` on top of `source-academy:master `?

**Comment by Anonymxtrix** [comment]
Hi @violinyap can we also include a sample _Source_ program here to test the modules under the "How Has This Been Tested?" section? Something like [this](https://github.com/source-academy/modules/pull/41).

**Comment by martin-henz** [comment]
> Hi @violinyap can we also include a sample _Source_ program here to test the modules under the "How Has This Been Tested?" section? Something like [this](https://github.com/source-academy/modules/pull/41).

@Anonymxtrix and @violinyap let's use the source-programs repository for example programs. There is a Modules demo section there, and I'm asking the teams to include their examples there.

**Comment by violinyap** [comment]
> > Hi @violinyap can we also include a sample _Source_ program here to test the modules under the "How Has This Been Tested?" section? Something like [this](https://github.com/source-academy/modules/pull/41).
> 
> @Anonymxtrix and @violinyap let's use the source-programs repository for example programs. There is a Modules demo section there, and I'm asking the teams to include their examples there.

Yes I will add the test program there, and link it in README file later

File: src/tabs/CopyGc/index.tsx [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+                      {item && item.length > 0
+                        ? // eslint-disable-next-line @typescript-eslint/no-unused-vars, arrow-body-style
+                          item.map((content, column) => {
+                            const color = this.getMemoryColor(content);
+                            const bgColor = this.getBackgroundColor(content);
+                            return (
+                              // eslint-disable-next-line jsx-a11y/mouse-events-have-key-events
+                              <div
+                                style={{ width: 14, backgroundColor: bgColor }}
+                                // onMouseOver={() => this.changeContent(content)}
```
Comment (Anonymxtrix) [inline src/tabs/CopyGc/index.tsx:392]: Consider removing unused code if possible

File: src/tabs/CopyGc/index.tsx [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+                          row * this.state.column + state.memorySize / 2
+                        }
+                      </span>
+                      {item && item.length > 0
+                        ? // eslint-disable-next-line @typescript-eslint/no-unused-vars, arrow-body-style
+                          item.map((content, column) => {
+                            const color = this.getMemoryColor(content);
+                            const bgColor = this.getBackgroundColor(content);
+                            return (
+                              // eslint-disable-next-line jsx-a11y/mouse-events-have-key-events
```
Comment (Anonymxtrix) [inline src/tabs/CopyGc/index.tsx:389]: No mouse events here

File: src/tabs/CopyGc/index.tsx [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+                    <div style={{ display: 'flex', flexDirection: 'row' }}>
+                      <span style={{ width: 30 }}>
+                        {
+                          // eslint-disable-next-line react/destructuring-assignment
+                          row * this.state.column + state.memorySize / 2
+                        }
+                      </span>
+                      {item && item.length > 0
+                        ? // eslint-disable-next-line @typescript-eslint/no-unused-vars, arrow-body-style
+                          item.map((content, column) => {
```
Comment (Anonymxtrix) [inline src/tabs/CopyGc/index.tsx:385]: Dont need `// eslint-disable-next-line @typescript-eslint/no-unused-vars, arrow-body-style` if you remove the column parameter from the lambda function.

File: src/tabs/CopyGc/index.tsx [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+                              <div
+                                style={{ width: 14, backgroundColor: bgColor }}
+                                // onMouseOver={() => this.changeContent(content)}
+                              >
+                                <canvas
+                                  width={10}
+                                  height={10}
+                                  style={{
+                                    backgroundColor: color,
+                                    // margin: 2,
```
Comment (Anonymxtrix) [inline src/tabs/CopyGc/index.tsx:399]: Consider removing unused code if possible

File: src/tabs/CopyGc/style.tsx [hunk 5]
```diff
@@ -0,0 +1,11 @@
+export const THEME_COLOR = {
+  BLUE: 'lightblue',
+  PINK: 'salmon',
+  GREY: '#707070',
+  GREEN: '#42a870',
+  YELLOW: '#f0d60e',
+};
```
Comment (Anonymxtrix) [inline src/tabs/CopyGc/style.tsx:7]: Maybe can use enum here as well, up to u though :\

File: src/tabs/MarkSweep/index.tsx [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+  leftDesc: String;
+  rightDesc: String;
+  unmarked: number;
+  marked: number;
+  queue: number[];
+};
+
+const MARK_SLOT = 1;
+class MarkSweep extends React.Component<Props, State> {
+  constructor(props: any) {
```
Comment (Anonymxtrix) [inline src/tabs/MarkSweep/index.tsx:37]: Consider using type `Props` defined in line 6

File: src/tabs/MarkSweep/index.tsx [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+}
+
+export default {
+  toSpawn: () => true,
+  body: (debuggerContext: any) => (
+    <MarkSweep debuggerContext={debuggerContext} />
+  ),
+  label: 'Mark Sweep Garbage Collector',
+  iconName: 'heat-grid',
+};
```
Comment (Anonymxtrix) [inline src/tabs/MarkSweep/index.tsx:460]: Similar issues as `CopyGc/index.tsx`

File: src/tabs/CopyGc/index.tsx [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+      flips: [0],
+      toMemoryMatrix: [],
+      fromMemoryMatrix: [],
+      firstChild: -1,
+      lastChild: -1,
+      command: '',
+      description: '',
+      rightDesc: '',
+      leftDesc: '',
+    };
```
[the remark this answers is not in the document]
Comment (violinyap) [reply src/tabs/CopyGc/index.tsx:62]: It is changed through the Slider that's why I'm saving it in the state.

File: src/bundles/copy_gc/index.ts [hunk 9]
```diff
@@ -88,14 +64,15 @@ function generateMemory(): void {
     desc: 'Memory initially empty.',
     leftDesc: '',
     rightDesc: '',
+    scan: -1,
+    free: -1,
   };
 
   commandHeap.push(obj);
```
Comment (Anonymxtrix) [inline src/bundles/copy_gc/index.ts:71]: By directly defining the `obj` inside the `commandHeap.push()` function, you eliminate the need to pollute the scope's name space and specify the type of `obj` to be `CommandHeapObject` in line 55. This is not absolutely necessary though, I think its completely up to you.

File: src/bundles/mark_sweep/index.ts [hunk 10]
```diff
@@ -184,34 +128,16 @@ function newCommand(
...
-    left,
-    -1,
-    newSizeLeft,
-    0,
-    heap,
-    desc,
-    'freed node',
-    ''
-  );
+  newCommand('SWEEP', left, -1, newSizeLeft, 0, heap, desc, 'freed node', '');
```
Comment (Anonymxtrix) [inline src/bundles/mark_sweep/index.ts:133]: Try not to add magic strings here unless it's absolutely necessary. Again, you can consider using TypeScript's enum.

File: src/bundles/mark_sweep/index.ts [hunk 11]
```diff
@@ -184,34 +128,16 @@ function newCommand(
...
 }
 
 function newMark(left, heap, queue): void {
-  const toSpace = 0;
-  const fromSpace = 0;
   const newSizeLeft = NODE_SIZE;
   const desc = `Marking node ${left} to be live memory`;
   newCommand(
     'MARK',
```
Comment (Anonymxtrix) [inline src/bundles/mark_sweep/index.ts:140]: Try not to add magic strings here unless it's absolutely necessary. Again, you can consider using TypeScript's enum.

File: src/bundles/mark_sweep/index.ts [hunk 12]
```diff
@@ -224,34 +150,28 @@ function newMark(left, heap, queue): void {
...
+  }
+  ROOTS = [];
 }
 
 function newUpdateSweep(right, heap): void {
-  const toSpace = 0;
-  const fromSpace = 0;
   const desc = `Set node ${right} to freelist`;
   newCommand(
     'SWEEP RESET',
```
Comment (Anonymxtrix) [inline src/bundles/mark_sweep/index.ts:174]: Try not to add magic strings here unless it's absolutely necessary. Again, you can consider using TypeScript's enum.

**Review by Anonymxtrix — APPROVED** [review]
Thanks for the updates! :)

**Pull request merged**